### ISSUE TITLE
Multi-host distribution

### ISSUE THREAD

**Issue by EiffL** [eelregit/pmwd#3]
In fantastic news, after over a year of waiting and checking every few months if it was working yet, it looks like finally it's possible to instantiate a distributed XLA runtime in Jax, which means.... Native access to NCCL collectives and composable parallelisation with pmap and xmap!!!

Demo of how to allocate 16 GPUs accross 4 nodes on Perlmutter here: https://github.com/EiffL/jax-gpu-cluster-demo

I'll be testing these things out and documenting my finding in this issue. Maybe won't be directly useful at first but at some point down the line we want to be able to run very large sims easily.

**Comment by EiffL** [comment]
@eelregit Making progress on this ^^ 
jaxdecomp is now able to do forward and backward FFTs https://github.com/DifferentiableUniverseInitiative/jaxDecomp 

Still have to add a few things, but really not far away from being usable as part of pmwd

**Comment by EiffL** [comment]
Ok, I've added halo exchange and cleaned up the interface. Also added gradients of these operations. You can also select which backend you want to use, MPI, NCCL, or NVSHMEM. As far as I can tell, this should be strictly superior to the cufftMP library, although now that I know how to do these bindings, it would be even easier to use cufftMP. 

Here is how you do a 3D FFT distributed on many GPUs with the current version of the API:
```python
from mpi4py import MPI
comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

import jax
import jax.numpy as jnp
import jaxdecomp

# Initialise the library, and optionally selects a communication backend (defaults to NCCL)
jaxdecomp.init()
jaxdecomp.config.update('halo_comm_backend', jaxdecomp.HALO_COMM_MPI)
jaxdecomp.config.update('transpose_comm_backend', jaxdecomp.TRANSPOSE_COMM_MPI_A2A)

# Setup a processor mesh (should be same size as "size")
pdims= [2,4]
global_shape=[1024,1024,1024]

# Initialize an array with the expected gobal size
array = jax.random.normal(shape=[1024//pdims[1], 
                                 1024//pdims[0], 
                                 1024], 
            key=jax.random.PRNGKey(rank)).astype('complex64')

# Forward FFT, note that the output FFT is transposed
karray = jaxdecomp.pfft3d(array, 
                global_shape=global_shape, pdims=pdims)

# Reverse FFT
recarray = jaxdecomp.ipfft3d(karray, 
        global_shape=global_shape, pdims=pdims)
        
# Add halo regions to our array
padded_array = jnp.pad(array, [(32,32),(32,32),(32,32)])
# Perform a halo exchange
padded_array = jaxdecomp.halo_exchange(padded_array,
                                       halo_extents=(32,32,32),
                                       halo_periods=(True,True,True),
                                       pdims=pdims,
                                       global_shape=global_shape)
```

Compiling is unfortunately not 100% trivial because it depends a lot on the local environment of the cluster, so I haven't managed to fully automatize it yet....

**Comment by EiffL** [comment]
One 1024^3 FFT on 4 V100 GPUs... 0.5 ms :rofl:

**Comment by EiffL** [comment]
Annnnd 50ms for a 2048^3 FFT on 16 V100 GPUs on 2 nodes...  :exploding_head: (also tagging @modichirag )

**Comment by eelregit** [comment]
Thanks!

> One 1024^3 FFT on 4 V100 GPUs... 0.5 ms rofl

Wow. But doesn't 0.5 ms sound too fast, like faster than the memory bandwidth on 1 GPU?

> Annnnd 50ms for a 2048^3 FFT on 16 V100 GPUs on 2 nodes

Should we expect strong scaling here, in which case (a bit more than) 2x the 1024^3 timing?

**Comment by EiffL** [comment]
0.5ms does sound really fast, but the result of the FFT seems to be correct, so.... maybe?

I'm not 100% sure what scaling we should expect, as a function of message size, it's possible that the cost is not the same, as the backend might switch between different strategies. 

Also, interesting note, this is using the NCCL backend, if I use the MPI backend on this setting and hardware I get 6s. I guess it will be very hardware and problem dependent, but that's what's nice with cuDecomp it includes an autotuning tool, that will find the best distribution strategy and backend for given hardware and problem size (which I havent interfaced in jax yet, but is there at the C++ level)

**Comment by eelregit** [comment]
50ms seems to be comparable to what cuFFTMp showed in https://developer.nvidia.com/blog/multinode-multi-gpu-using-nvidia-cufftmp-ffts-at-scale/

For me one 1024^3 FFT on 1 A100 seems to take more than 30ms. Are you sure that 0.5ms on 4 V100 is not underestimating?

**Comment by EiffL** [comment]
oups ^^' you are right, I didnt include a block until ready... 
New timings: 
- 2048^3 on 16 GPUs using NCCL: 8s
- 1024^3 on 4 GPUs using NCCL: 0.2s

**Comment by eelregit** [comment]
Still very promising. I wonder if the difference in performance and scaling is mainly from the hardware (nvlink, nvswitch etc)?

**Comment by wendazhou** [comment]
Nvidia claims that they can achieve close to perfect weak scaling with cuFFTmp in the 2^30 elements / GPU range (up to about 4k GPU), but I know that that library leans heavily on `nvshmem` communication to achieve optimal overlapping. nvidia claims that their cluster can do 2048^3 in ~100ms on 16 GPUs (albeit A100), so it definitely might be worth looking into using that library directly / setting up the correct hardware config for cuFFTmp and cuDECOMP.

**Comment by EiffL** [comment]
Yep, it should be trivial to add an op for cufftmp in jaxdecomp as it's already part of the nvhpc SDK, so no need to compile an external library :-) I'm traveling this week, but if I catch a little quiet time I'll add the option to use cufftMP. Unless you want to have a go at it @wendazhou ;-) 

One thing though that I thought about, reading the documentation it looks like nvshmem memory needs to be allocated in a particular fashion, which is different from a standard cuda device memory. That means we can't directly use the input/output buffers allocated by XLA, and that the op will need to allocate its own buffers using nvshmem, that will kind of double the memory needed to run the FFT. 

In my current implementation, I do the transform in-place within the input/buffer allocated by XLA. I also need a workspace buffer, of size determined by cudecomp (I think something like twice the size of the input array) , also allocated by XLA.

**Comment by wendazhou** [comment]
I'm also travelling for NeurIPS this week, probably won't have time to look at it. For the memory, I don't think that input / output buffers need to be allocated using special functions, only the scratch memory itself, which `cuFFTmp` handles internally (but this indeed requires linear memory in the transform size, see [doc](https://docs.nvidia.com/hpc-sdk/cufftmp/usage/memory.html)).

I think the main work in addition to plumbing everything together will be to figure out how to describe the data layout correctly [doc](https://docs.nvidia.com/hpc-sdk/cufftmp/usage/api_usage.html#usage-with-custom-slabs-and-pencils-data-decompositions)

**Comment by eelregit** [comment]
@hai4john

**Comment by EiffL** [comment]
@eelregit ... ok, took about a year, but it's working nicely with the latest version of JAX thanks to the heroic efforts of my collaborator @ASKabalan :-) 

We have updated jaxDecomp https://github.com/DifferentiableUniverseInitiative/jaxDecomp to be compatible with the native distribution, and we have a rough prototype of a distributed LPT demo here: https://github.com/DifferentiableUniverseInitiative/JaxPM/issues/19#issuecomment-2103839585 

We have been able to scale it to 24 GPUs (the max I can allocated at Flatiron), and to give you an idea, it executes an LPT simulation (no PM iteration) on a 2048^3 mesh in 4.7s. We haven't yet carefully profiled the execution, it's not impossible we could do even better, but at least it means we can do reasonable cosmological volumes in a matter of seconds.

With @ASKabalan we are probably going to integrate this in JaxPM, as a way to prototype user APIs compatible with distribution (it's not completely trivial, we don't want to bother users with a lot of stuff if they are running on a single GPU setting). 
But I still have in mind to push distribution to pmwd, so wanted to check-in with you to see what you think and if you have another path to distribution in mind already.

**Comment by eelregit** [comment]
Thanks @EiffL for letting me know. Let's find a time (maybe next week) to chat via email?

**Comment by EiffL** [comment]
@eelregit here is a minimal demo of LPT implemented using jaxdecomp: https://github.com/DifferentiableUniverseInitiative/jaxDecomp/blob/main/examples/lpt_nbody_demo.py

**Comment by EiffL** [comment]
@eelregit timings  https://flanusse.net/talks/Split2024/#/15/0/1
